### ISSUE TITLE
Opening files associated with a GLFW program after launch

### ISSUE THREAD

**Issue by Andos** [glfw/glfw#1024]
Normally file associations with GLFW apps is a topic outside the scope of GLFW itself - but in this case the design of GLFW actively prevents this from happening on macOS:

On macOS the application's needs this method:
- (BOOL)application:(NSApplication *)sender openFile:(NSString *)filename;
to be able to open both files from application launch and after the application has launched.

I propose a simple macOS specific callback that is called when this application delegate method is called and then returns the answer to the OS.

**Comment by fogleman** [comment]
Yes, please. Or it could invoke the existing path drop callback, although it's probably best if you can differentiate the two types of events: http://www.glfw.org/docs/latest/input_guide.html#path_drop

Can you think of any workaround in the meantime @elmindreda?

**Comment by fogleman** [comment]
Worth noting the similar openFiles method:

https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428742-application?language=objc

**Comment by fogleman** [comment]
I threw in a quick hack locally, in `cocoa_window.m`:

```
- (BOOL)application:(NSApplication *)sender openFile:(NSString *)filename
{
    _GLFWwindow* window = _glfw.windowListHead;
    if (window == nil) {
        return NO;
    }
    char** paths = calloc(1, sizeof(char*));
    paths[0] = strdup([filename UTF8String]);
    _glfwInputDrop(window, 1, (const char**)paths);
    free(paths[0]);
    free(paths);
    return YES;
}
```

But this code is invoked before I even have time to call `glfwSetDropCallback`. It might be necessary to save a copy of the filenames and have a new method that lets the app retrieve them?

**Comment by andreyugolnik** [comment]
Any ideas when it ready?

In my open source image viewer for Linux and macOS I need support of
`- (BOOL)application:(NSApplication *)sender openFile:(NSString *)filename;`

**Comment by elmindreda** [comment]
Sorry for the late reply.

I really want to add support for this but I also really don't want to add platform-specific callbacks if I can possibly avoid it.

> It might be necessary to save a copy of the filenames and have a new method that lets the app retrieve them?

Yes, that's one possibility.

Another would be to save them as above but pass them to the drop callback once there is a [main window](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/WinPanel/Concepts/ChangingMainKeyWindow.html#//apple_ref/doc/uid/20000236-129193). In this particular case it may even be an advantage that the current drop callback lacks cursor position data.

**Comment by andreyugolnik** [comment]
I think good idea to allow inject own application delegate.

**Comment by fogleman** [comment]
@elmindreda Any new ideas for this issue? :)

**Comment by bear24rw** [comment]
Right now I spawn a GCD thread which pools for the drop callback to be registered and then calls it. It's super hacky so I'd love if there was official support for this.

**Comment by qno** [comment]
@elmindreda - is there any progress regarding this topic? How does the plan looks like to solve this issue?

Best Regards

**Comment by kritzikratzi** [comment]
afaik this doesn't need a fork, it should be possible to solve it with swizzling

**Comment by kritzikratzi** [comment]
ok, so for all the people looking around, here is a workaround that needs no changes to glfw at all. 

the code is based more or less directly on this tutorial https://nshipster.com/method-swizzling/

`GLFWCustomDelegate.h`

	#import <Foundation/Foundation.h>
	#import <AppKit/AppKit.h>

	NS_ASSUME_NONNULL_BEGIN

	@interface GLFWCustomDelegate : NSObject
	+ (void)load; // load is called before even main() is run (as part of objc class registration)
	@end

	NS_ASSUME_NONNULL_END


`GLFWCustomDelegate.m`

	#import "GLFWCustomDelegate.h"
	#import <objc/runtime.h>

	// part of your application
	extern void forward_load_message(const char * filename); 

	@implementation GLFWCustomDelegate

	+ (void)load{
		static dispatch_once_t onceToken;
		dispatch_once(&onceToken, ^{
			Class class = objc_getClass("GLFWApplicationDelegate");
		
			[GLFWCustomDelegate swizzle:class src:@selector(application:openFile:) tgt:@selector(swz_application:openFile:)];
			[GLFWCustomDelegate swizzle:class src:@selector(application:openFiles:) tgt:@selector(swz_application:openFiles:)];
		});
	}

	+ (void) swizzle:(Class) original_c src:(SEL)original_s tgt:(SEL)target_s{
		Class target_c = [GLFWCustomDelegate class];
		Method originalMethod = class_getInstanceMethod(original_c, original_s);
		Method swizzledMethod = class_getInstanceMethod(target_c, target_s);
	
		BOOL didAddMethod =
		class_addMethod(original_c,
						original_s,
						method_getImplementation(swizzledMethod),
						method_getTypeEncoding(swizzledMethod));
	
		if (didAddMethod) {
			class_replaceMethod(original_c,
								target_s,
								method_getImplementation(originalMethod),
								method_getTypeEncoding(originalMethod));
		} else {
			method_exchangeImplementations(originalMethod, swizzledMethod);
		}
	}

	- (BOOL)swz_application:(NSApplication *)sender openFile:(NSString *)filename{
		forward_load_message(filename.UTF8String);
	}

	- (void)swz_application:(NSApplication *)sender openFiles:(NSArray<NSString *> *)filenames{
		forward_load_message(filenames.firstObject.UTF8String);
	}

	@end



`Customize your existing main.cpp`

	// file: main.cpp
	// this is mostly pseudocode, adjust to your needs
	CoolApp * myApp = nullptr; 
	bool startup_kind_of_done = false;
	string startup_file_to_load = "";


	extern "C"{
		void forward_load_message(const char * filename){
			if(!startup_kind_of_done){
				// this is the case where a user double clicks a file, but your app is not yet open
				startup_file_to_load = filename;
			}
			else{
				myApp->load(string(filename));
			}
		}
	}


	int main(){
		// 1. set up windows, or whatever you need to do ... 
		myApp = new CoolApp(); 

		// 2. check if we should load a file
		startup_kind_of_done = true; 
		if(start_file_to_load != "" ){
			myApp->loadFile(startup_file_to_load); 
		}
	
		// 3. we are done
		runSomeKindOfEventLoop(myApp); 
	}

**Comment by ws909** [comment]
Is there no similar API for this in Windows and Linux? I have been looking for a way to catch these events too, as I'd like to implement functionality similar to what I'm used to, on MacOS. On Windows, you can open files with specific applications, from the File Explorer. This is pretty much the same as MacOS offers. I would assume Linux has the same. So it is unclear to me why this would have to be a MacOS-only feature.

**Comment by kritzikratzi** [comment]
@ws909 iirc windows and linux handle this differently - they pass the file as the second argument to main/winmain and you can catch it there. on osx you have to catch it in `application:(NSApplication *)sender openFile` and it won't be in your args array. 

the even bigger difference is that on osx only one instance of your executable will ever be launched and you have to take care of multi-document applications yourself, while on windows and linux a new instance will be launched every time you double click a file.

**Comment by ws909** [comment]
@kritzikratzi That clarifies. I still do believe such events are still such a basic part of an application, though, that it should be included in GLFW. I can think of two ways to implement cross-platform support:

1. Platform decides the model.
2. MacOS approach on all platforms

Java (as I use LWJGL):
```
void openFiles(String[] paths) {
    ...
}

static void main(String[] arguments) {

    glfwInitHint(GLFW_REDIRECT_APPLICATION, GLFW_FALSE); // Possibly GLFW_REDIRECT_OPEN_FILES instead
    glfwInitHint(GLFW_INIT_FILE_ARGUMENTS, arguments);

    glfwInit()
    
    glfwSetOpenFilesCallback(openFiles);

    final var paths = glfwGetOpenedFilenames();

    if (paths != null) {
        openFiles(paths); // Or any other way to respond
    }
    
    while (true) {
        glfwWaitEvents();
    }
}
```

This is model 1. The `GLFW_INIT_FILE_ARGUMENTS` is used to set the arguments that `glfwInit` parses to find which files to open. These can be queried with `glfwGetOpenedFilenames`. On MacOS, there won't be any file paths delivered through the arguments, however, the current implementation, #1579, solves this. On MacOS, when the system again invokes `NSAppDelegate::openFiles`, the `openFiles` method set as the callback, will be run. `glfwGetOpenedFilenames` will return these new files only. `NSAppDelegate::openFile` is ignored in this implementation. 

For model 2, set `GLFW_REDIRECT_APPLICATION` to `GLFW_TRUE`. First time the application is launched, on Linux and Windows, `glfwInit` sets this "instance" as the main one, for the user. Subsequent executions of the same application, will make `glfwInit` transfer the file paths to the main instance, which are forwarded to the callback. Then `glfwInit` aborts initilization, and returns `GLFW_FALSE`. Can check this with `glfwDidTransfer` or `glfwWasAlreadyLaunched`, etc.

**Comment by kritzikratzi** [comment]
do it :) everyone is gonna enjoy. i already have a solution that is not great, but it works for me (posted far above)

**Comment by ws909** [comment]
@kritzikratzi Might look into it later. :) Did look at your solution. Seems like a nice hack, but it's unnecessarily complicated for something that should be easily done by GLFW.

**Comment by kritzikratzi** [comment]
of course it's a hack :D but it works with any glfw version and it integrates easy as pie into a c/c++ project. i guess from a java perspective it's pita.
